File: jelenoid-server-go/internal/services/session_publisher.go
package services

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/balakshievas/jelenoid-server-go/internal/dto"
	"github.com/nats-io/nats.go"
)

type SessionPublisher struct {
	nc *nats.Conn
}

func NewSessionPublisher(nc *nats.Conn) *SessionPublisher {
	return &SessionPublisher{nc: nc}
}

func (p *SessionPublisher) publish(info *dto.SessionInfo) {
	if p.nc == nil {
		return
	}
	data, err := json.Marshal(info)
	if err != nil {
		return
	}
	if err := p.nc.Publish("sessions.events", data); err != nil {
		fmt.Printf("Failed to publish session event: %v\n", err)
	}
}

func (p *SessionPublisher) CreateSessionAndPublish(platform, version string) *dto.SessionInfo {
	now := time.Now()
	info := &dto.SessionInfo{
		ID:        generateUUID(),
		StartTime: &now,
		Platform:  platform,
		Version:   version,
		Status:    "started",
	}
	p.publish(info)
	return info
}

func (p *SessionPublisher) EndSessionByRemoteAndPublish(info *dto.SessionInfo) *dto.SessionInfo {
	if info == nil {
		return nil
	}
	now := time.Now()
	info.EndTime = &now
	info.Status = "finished"
	info.EndedBy = "by remote"
	p.publish(info)
	return info
}

func (p *SessionPublisher) ErrorSessionAndPublish(info *dto.SessionInfo) *dto.SessionInfo {
	if info == nil {
		return nil
	}
	now := time.Now()
	info.EndTime = &now
	info.Status = "error"
	info.EndedBy = "error"
	p.publish(info)
	return info
}

func (p *SessionPublisher) EndInactiveSessionAndPublish(info *dto.SessionInfo) *dto.SessionInfo {
	if info == nil {
		return nil
	}
	now := time.Now()
	info.EndTime = &now
	info.Status = "finished"
	info.EndedBy = "by inactive"
	p.publish(info)
	return info
}

func (p *SessionPublisher) CleanupSessionAndPublish(info *dto.SessionInfo) *dto.SessionInfo {
	if info == nil {
		return nil
	}
	now := time.Now()
	info.EndTime = &now
	info.Status = "finished"
	info.EndedBy = "cleanup"
	p.publish(info)
	return info
}

func generateUUID() string {
	b := make([]byte, 16)
	for i := range b {
		b[i] = byte(i)
	}
	return fmt.Sprintf("%x", b)
}

File: legacy/jelenoid-server-java/src/main/java/com/balakshievas/jelenoid/config/SessionPublisher.java
package com.balakshievas.jelenoid.config;

import com.balakshievas.jelenoid.dto.SessionInfo;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import io.nats.client.Connection;
import lombok.extern.slf4j.Slf4j;

import java.time.LocalDateTime;
import java.util.UUID;

@Slf4j
public class SessionPublisher {

    private static final String SUBJECT = "sessions.events";
    private final Connection nats;

    public SessionPublisher(Connection nats) {
        this.nats = nats;
    }

    public void publish(SessionInfo sessionInfo) {
        if (nats == null) {
            return;
        }
        try {
            byte[] json = new ObjectMapper()
                    .registerModule(new JavaTimeModule())
                    .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
                    .writeValueAsBytes(sessionInfo);
            nats.publish(SUBJECT, json);
        } catch (Exception ignored) {
        }
    }

    public SessionInfo createSessionAndPublish(String platform, String version) {

        SessionInfo sessionInfo = new SessionInfo(
                UUID.randomUUID().toString(),
                LocalDateTime.now(),
                null,
                platform,
                version,
                "started",
                null
        );
        publish(sessionInfo);
        return sessionInfo;
    }

    public SessionInfo endSessionByRemoteAndPublish(SessionInfo sessionInfo) {
        if(sessionInfo == null) {
            return null;
        }
        sessionInfo.setEndTime(LocalDateTime.now());
        sessionInfo.setStatus("finished");
        sessionInfo.setEndedBy("by remote");
        publish(sessionInfo);
        return sessionInfo;
    }

    public SessionInfo errorSessionAndPublish(SessionInfo sessionInfo) {
        if(sessionInfo == null) {
            return null;
        }
        sessionInfo.setEndTime(LocalDateTime.now());
        sessionInfo.setStatus("error");
        sessionInfo.setEndedBy("error");
        publish(sessionInfo);
        return sessionInfo;
    }

    public SessionInfo endInactiveSessionAndPublish(SessionInfo sessionInfo) {
        if(sessionInfo == null) {
            return null;
        }
        sessionInfo.setEndTime(LocalDateTime.now());
        sessionInfo.setStatus("finished");
        sessionInfo.setEndedBy("by inactive");
        publish(sessionInfo);
        return sessionInfo;
    }

    public SessionInfo cleanupSessionAndPublish(SessionInfo sessionInfo) {
        if(sessionInfo == null) {
            return null;
        }
        sessionInfo.setEndTime(LocalDateTime.now());
        sessionInfo.setStatus("finished");
        sessionInfo.setEndedBy("cleanup");
        publish(sessionInfo);
        return sessionInfo;
    }
}
